crates/sim2h/src/lib.rs

```diff
@@ -684,128 +746,330 @@ impl Sim2hState {
         entry_loc: Location,
         redundant_count: u64,
     ) -> Vec<(AgentId, AgentInfo)> {
+        let _m = self.metric_gen.timer("sim2h-state-agents_in_neighbourhood");
         self.get_space(&space)
             .agents_supposed_to_hold_entry(entry_loc, redundant_count)
             .into_iter()
             .collect::<Vec<(AgentId, AgentInfo)>>()
     }
 }
 
-enum PoolTask {
-    //    VerifyPayload(Result<(Lib3hUri, WireMessage, AgentPubKey), ()>),
+#[derive(Debug)]
+struct Sim2hComHandleMessage {
+    uri: Lib3hUri,
+    message: WireMessage,
+    signer: AgentId,
+}
+
+#[derive(Debug)]
+struct Sim2hComHandleJoined {
+    uri: Lib3hUri,
+    space_address: SpaceHash,
+    agent_id: AgentId,
+    message: WireMessage,
+}
+
+#[derive(Debug)]
+enum Sim2hCom {
+    HandleMessage(Box<Sim2hComHandleMessage>),
+    HandleJoined(Box<Sim2hComHandleJoined>),
     Disconnect(Vec<Lib3hUri>),
 }
 
+#[derive(Clone)]
+/// A clonable reference to our Sim2h instance that can be passed
+/// into `'static` async blocks && still be able to make sim2h calls
+struct Sim2hHandle {
+    state: Arc<tokio::sync::Mutex<Sim2hState>>,
+    send_com: tokio::sync::mpsc::UnboundedSender<Sim2hCom>,
+    dht_algorithm: DhtAlgorithm,
+    metric_gen: MetricsTimerGenerator,
+}
+
+impl Sim2hHandle {
+    pub fn new(
+        state: Arc<tokio::sync::Mutex<Sim2hState>>,
+        send_com: tokio::sync::mpsc::UnboundedSender<Sim2hCom>,
+        dht_algorithm: DhtAlgorithm,
+        metric_gen: MetricsTimerGenerator,
+    ) -> Self {
+        Self {
+            state,
+            send_com,
+            dht_algorithm,
+            metric_gen,
+        }
+    }
+
+    /// generate a new metrics timer
+    pub fn metric_timer(&self, tag: &'static str) -> MetricsTimer {
+        self.metric_gen.timer(tag)
+    }
+
+    /// get our current dht algorithm
+    pub fn dht_algorithm(&self) -> &DhtAlgorithm {
+        &self.dht_algorithm
+    }
+
+    /// acquire a mutex lock to our state data
+    pub async fn lock_state(&self) -> tokio::sync::MutexGuard<'_, Sim2hState> {
```

Comment:
this is a little weird syntactically - since this itself is an async function and we're awaiting within it, it works out that the calling function will await on the lock being acquired.

It's true you need to be careful that that guard goes out of scope before any more awaits in the using function, but that's just something you have to be careful about in any rust code involving futures...

Reply:
sorry -- re-read your comment, your concern is about not doing this in the first place so you don't have to be careful about it.

Agree that would be preferable - I was super careful about all the current uses making sure they are dropped appropriately and no `await`s happen in the mean time.

If the IM state stuff works out how I'm hoping I'll be able to get rid of this function / these MutexGuards entirely.

Reply:
Yes, agree with the last statement. As per discussion the stand-up today, this function will go away when implementing the IM/actor/internal mutability stuff.

Ok, fine with merging this for now.